File: src/typescript/countingSort.ts
const countingSort = (seq: number[]): void => {
  let maxValue = seq[0];
  let minValue = seq[0];
  for (let i = 1; i < seq.length; i++) {
    if (maxValue < seq[i]) {
      maxValue = seq[i];
    }
    if (minValue > seq[i]) {
      minValue = seq[i];
    }
  }

  let counts = new Array(maxValue - minValue + 1).fill(0);
  for (let i = 0; i < seq.length; i++) {
    counts[i + minValue] += 1;
  }

  let temp = 0
  for (let i = 0; i < counts.length; i++) {
    for (let j = 0; j < counts[i]; j++) {
      seq[temp] = i + minValue;
      temp += 1;
    }
  }
}


const countingSortTest = () => {
  let seq1 = [16, 7, 9, 5, 65, 49, 37, 3, 28, 2, 21, 12, 4];
  let seq1Sorted = Object.assign([], seq1);
  countingSort(seq1Sorted);

  console.log(`seq1: ${seq1}\nseq1Sorted: ${seq1Sorted}`);
}

countingSortTest();

File: src/rust/counting_sort.rs
pub fn counting_sort(seq: &mut Vec<i32>) {
    let mut max_value = seq[0];
    let mut min_value = seq[0];
    for i in 1..seq.len() {
        if max_value < seq[i] {
            max_value = seq[i];
        }
        if min_value > seq[i] {
            min_value = seq[i];
        }
    }

    let mut counts = vec![0; ((max_value - min_value) as usize + 1) as usize];
    for i in 0..seq.len() {
        counts[(seq[i] - min_value) as usize] += 1;
    }

    let mut temp = 0;
    for i in 0..counts.len() {
        for _ in 0..counts[i] {
            seq[temp] = (i + min_value as usize) as i32;
            temp += 1;
        }
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    fn get_vector_i32() -> Vec<i32> {
        vec![16, 7, 9, 5, 65, 49, 37, 3, 28, 2, 21, 12, 4]
    }


    fn is_sorted<T: PartialOrd>(seq: &[T]) -> bool {
        let mut prev = &seq[0];
        for elem in seq.iter().skip(1) {
            if prev > elem {
                return false;
            }
            prev = elem;
        }
        true
    }

    #[test]
    fn test_sanity() {
        let mut v_i32 = get_vector_i32();
        counting_sort(&mut v_i32);
        assert!(is_sorted(&v_i32));
    }
}
